internal/devconfig/ast.go

```diff
@@ -0,0 +1,282 @@
+// saving changes.
+//
+//   - Unmarshalling is still done with encoding/json.
+//   - Marshalling is done by calling configAST.root.Pack to encode the AST as
+//     hujson/JWCC. Therefore, any changes to a Config struct will NOT
+//     automatically be marshaled back to JSON. Support for modifying a part of
+//     the JSON must be explicitly implemented in configAST.
+//   - Validation with the AST is complex, so it doesn't do any. It will happily
+//     append duplicate object keys and panic on invalid types. The higher-level
+//     Config type is responsible for tracking state and making valid edits to
+//     the AST.
+//
+// Be aware that there are 4 ways of representing a package in devbox.json that
+// the AST needs to handle:
+//
+//  1. ["name"] or ["name@version"] (versioned name array)
+//  2. {"name": "version"} (packages object member with version string)
+//  3. {"name": {"version": "1.2.3"}} (packages object member with package object)
+//  4. {"github:F1bonacc1/process-compose/v0.40.2": {}} (packages object member with flakeref)
+type configAST struct {
+	root hujson.Value
+}
+
+// parseConfig parses the bytes of a devbox.json and returns a syntax tree.
+func parseConfig(b []byte) (*configAST, error) {
+	root, err := hujson.Parse(b)
+	if err != nil {
+		return nil, err
+	}
+	return &configAST{root: root}, nil
+}
+
+// packagesField gets the "packages" field, initializing it if necessary. The
+// member value will either be an array of strings or an object. When it's an
+// object, the keys will always be package names and the values will be a
+// string or another object. Examples are:
+//
+//   - {"packages": ["go", "hello"]}
+//   - {"packages": {"go": "1.20", "hello: {"platforms": ["aarch64-darwin"]}}}
+//
+// When migrate is true, the packages value will be migrated from the legacy
+// array format to the object format. For example, the array:
+//
+//	["go@latest", "hello"]
+//
+// will become:
+//
+//	{
+//		"go": "latest",
+//		"hello": ""
+//	}
+func (c *configAST) packagesField(migrate bool) *hujson.ObjectMember {
+	rootObject := c.root.Value.(*hujson.Object)
+	i := c.memberIndex(rootObject, "packages")
+	if i != -1 {
+		switch rootObject.Members[i].Value.Value.Kind() {
+		case '[':
+			if migrate {
+				c.migratePackagesArray(&rootObject.Members[i].Value)
+				c.root.Format()
+			}
+		case 'n':
+			// Initialize a null packages field to an empty object.
+			rootObject.Members[i].Value.Value = &hujson.Object{
+				AfterExtra: []byte{'\n'},
+			}
+			c.root.Format()
+		}
+		return &rootObject.Members[i]
+	}
+
+	// Add a packages field to the root config object and initialize it with
+	// an empty object.
+	rootObject.Members = append(rootObject.Members, hujson.ObjectMember{
+		Name: hujson.Value{
+			Value:       hujson.String("packages"),
+			BeforeExtra: []byte{'\n'},
+		},
+		Value: hujson.Value{Value: &hujson.Object{}},
+	})
+	c.root.Format()
```

Comment:
what does Format do?

Reply:
Generally applies indentation and makes things look nice. If the input has any commas or comments (making it JWCC and not JSON), it also adds trailing commas and aligns fields.